### PR TITLE
Use `Reset()` in default constructor of `Memory<T>`.

### PR DESCRIPTION
**Pull request by YohannDudouit**
| PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2358](https://github.com/mfem/mfem/pull/2358) | @YohannDudouit | @tzanio | @camierjs + @pazner + @v-dobrev + @acfisher | 06/26/21 | 11/10/21 | 11/11/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: linalg/vector.hpp
Comment (pazner) [inline linalg/vector.hpp:70]: Maybe add a comment here that the default constructor for `data` will zero-out all the `Memory` metadata and host pointer? (It says it sets `data = NULL`, but that happens only implicitly).

File: general/mem_manager.hpp
Comment (pazner) [inline general/mem_manager.hpp:172]: Better comment here would be useful.

**Review by pazner — COMMENTED** [review]
I think this is an improvement — it's usually the desired behavior, and the only downside is the couple of operations it takes to zero out the pointer and flags. Uninitialized-by-default variables are widely known to be the cause of many bugs, and so I think this behavior is safer.

I suspect there are virtually no performance implications, but (cf. [this comment](https://github.com/mfem/mfem/pull/2357#pullrequestreview-693561838)) I would be interested in having a way to automatically identify performance regressions.

File: linalg/vector.hpp
Comment (YohannDudouit) [reply linalg/vector.hpp]: Tell me if you like it more.

File: general/mem_manager.hpp
Comment (YohannDudouit) [reply general/mem_manager.hpp]: Tell me if you like it more.

Comment (pazner) [reply general/mem_manager.hpp]: Looks great

File: linalg/vector.hpp
Comment (pazner) [reply linalg/vector.hpp]: Looks great

**Comment by v-dobrev** [comment]
This change adds another difference in the `Memory` class behavior compared to a pointer -- it will be good to mention this in the description of the class -- unlike a pointer, it is initialized as empty (equivalent to setting a pointer to NULL) by the default ctor.

One drawback of this change is that some code will now perform redundant initialization, e.g. `Vector::Vector(int s)` in the case `s > 0` -- it will first initialize `data` with `data.Reset` and then re-initialize with `data.New`. If we could conditionally call different `Memory` constructors, this would not have been a problem here, however we cannot.

Personally, I prefer to stick to the current simple pointer-like behavior, since this is mostly internal, low-level class that users will rarely, if ever, use directly -- user code should just use class `Array` instead. With this behavior we can fine-tune things (like the above ctor) and be more explicit about creating/initializing/(shallow)copying/overwriting/deleting `Memory` objects.

Obviously, I'm biased here, since I intentionally designed this class to be the way it is. I'll let other reviewers approve the PR if they like it -- I'll neither approve, nor reject it.

**Comment by YohannDudouit** [comment]
> This change adds another difference in the `Memory` class behavior compared to a pointer -- it will be good to mention this in the description of the class -- unlike a pointer, it is initialized as empty (equivalent to setting a pointer to NULL) by the default ctor.

I can add this, though I don't think we should push too much this comparison with pointers as this class really does not work like a pointer.
> 
> One drawback of this change is that some code will now perform redundant initialization, e.g. `Vector::Vector(int s)` in the case `s > 0` -- it will first initialize `data` with `data.Reset` and then re-initialize with `data.New`. If we could conditionally call different `Memory` constructors, this would not have been a problem here, however we cannot.

I'm planning to not call `New` but the constructor `Memory::Memory(int)` instead (which would avoid the call to `Reset()`), currently `Memory(0)` and `New(0)` are broken (we would expect `New(0)` to do relatively the same thing as `Reset()` but it doesn't). I started this in #2353 but I had some issues, so I favored a more atomic and incremental approach.
Currently, one has to write the following code to initialize a `Memory` object:
```c++
size > 0 ? data.New(size) : data.Reset();
```
This is really asking developers unnecessary efforts, do you really expect everyone that ever use a `Memory` object to be aware that `New(0)` should not be used, but `Reset()` instead? Handling the `0` case inside `New` is much cleaner IMHO:
```c++
template <typename T>
inline void Memory<T>::New(int size)
{
   if (size > 0)
   {
      capacity = size;
      flags = OWNS_HOST | VALID_HOST;
      h_mt = MemoryManager::GetHostMemoryType();
      h_ptr = (h_mt == MemoryType::HOST) ? NewHOST(size) :
              (T*)MemoryManager::New_(nullptr, size*sizeof(T), h_mt, flags);
   }
   else
   {
      Reset();
   }
}
```
> 
> Personally, I prefer to stick to the current simple pointer-like behavior, since this is mostly internal, low-level class that users will rarely, if ever, use directly -- user code should just use class `Array` instead. With this behavior we can fine-tune things (like the above ctor) and be more explicit about creating/initializing/(shallow)copying/overwriting/deleting `Memory` objects.

If you were the only one using this class that would probably be okay, but we're a team and it's difficult to use this class without creating bugs. I think anything that makes this class safer and cleaner is good to take, the less we have to worry about implicit rules not enforced by the code the more we can be productive and focus on writing applied math algorithms. I really don't understand why you don't want to make the code safer and cleaner when this has no performance impact.
> 
> Obviously, I'm biased here, since I intentionally designed this class to be the way it is. I'll let other reviewers approve the PR if they like it -- I'll neither approve, nor reject it.

If you have any example of code that would show any advantage of the current design over this one, please share with the team.

**Comment by pazner** [comment]
> This change adds another difference in the `Memory` class behavior compared to a pointer -- it will be good to mention this in the description of the class -- unlike a pointer, it is initialized as empty (equivalent to setting a pointer to NULL) by the default ctor.

I think Veselin brings up some valid points, but I wanted to mention one (somewhat pedantic) thing: the default constructor for a pointer *does* set it to NULL:

```c++
double *p{}; // default constructor
p == nullptr; // true
```

Now admittedly, C++ is a little weird about this, since usually you don't need to add `{}` to call the default constructor, but that is a weirdness of the language: *default initialization* of class types calls the constructor, and *default initialization* of non-class types does *not* call the constructor.

Another example,
```c++
std::vector<double*> v(1); // one-element default-initialized vector
v[0] == nullptr; // true
```

So it is actually *consistent* with the behavior of a pointer to set it NULL in the default constructor. What is inconsistent is that a raw pointer is a non-class type, and `Memory<T>` is a class type, but that is not in our control.

**Comment by v-dobrev** [comment]
Adding `()` after the name of a variable is called value-initialization (https://en.cppreference.com/w/cpp/language/value_initialization) which is different from default-initialization (https://en.cppreference.com/w/cpp/language/default_initialization). These are two different things, at least for simple types like `int`, `double`, and pointers.

Notably, STL containers invoke value-initialization when constructing new elements.

**Comment by pazner** [comment]
> Adding `()` after the name of a variable is called value-initialization (https://en.cppreference.com/w/cpp/language/value_initialization) which is different from default-initialization (https://en.cppreference.com/w/cpp/language/default_initialization). These are two different things, at least for simple types like `int`, `double`, and pointers.
> 
> Notably, STL containers invoke value-initialization when constructing new elements.

Yes, I know. The point being that *default initialization* of non-class types does *not* call the default constructor. The default constructor of raw pointers sets the pointer to NULL.

**Comment by v-dobrev** [comment]
> The point being that _default initialization_ of non-class types does _not_ call the default constructor. The default constructor of raw pointers sets the pointer to NULL.

I'm not sure "default constructor" is actually defined for pointers -- neither `int *p` nor `int *p()` invoke a "default constructor". According to https://en.cppreference.com/w/cpp/language/value_initialization, the second form performs zero initialization which in turn is done by the rule: "If T is a scalar type, the object's initial value is the integral constant zero explicitly converted to T. " -- it does not mention "default constructor".

In any case, this distinction is irrelevant for class `Memory` since both forms invoke the default constructor:
```c++
      Memory<int> m1;
      Memory<int> m2{};
      Memory<int> m3();  // cannot use: "warning: interpreted as function declaration"
```

**Comment by v-dobrev** [comment]
> This is really asking developers unnecessary efforts, do you really expect everyone that ever use a `Memory` object to be aware that `New(0)` should not be used, but `Reset()` instead? Handling the `0` case inside `New` is much cleaner IMHO:

The choice in class Vector to set the pointer to NULL when size is 0, is specific choice for that class. The choice for `Memory::New(0)` to behave as it does is for consistency with a `new` expression -- when invoked with a size of 0, it returns a non-null pointer. This choice seems more appropriate to me for class `Memory`.

**Comment by pazner** [comment]
> I'm not sure "default constructor" is actually defined for pointers -- neither `int *p` nor `int *p()` invoke a "default constructor". According to https://en.cppreference.com/w/cpp/language/value_initialization, the second form performs zero initialization which in turn is done by the rule: "If T is a scalar type, the object's initial value is the integral constant zero explicitly converted to T. " -- it does not mention "default constructor".

You're probably right, and "default constructor" may not be the precise terminology in the standard.

What I was trying to communicate in the above posts is the following:
```c++
double *p1; // implementation-defined value
double *p2{}; // nullptr

Memory<double> m1; // (mfem master) implementation-defined value
Memory<double> m2{}; // (mfem master) implementation-defined value

Memory<double> m1; // (this PR) nullptr
Memory<double> m2{}; // (this PR) nullptr
```

In master, `Memory` is inconsistent with a raw pointer for value initialization, and in this PR `Memory` is inconsistent with a raw pointer for default initialization. There is no way (as far as I know) to make `Memory` consistent in both cases, because a class type behaves differently from a non-class type. So, we have to pick which inconsistency we prefer.

**Comment by v-dobrev** [comment]
@pazner:
> So, we have to pick which inconsistency we prefer.

I agree, this specific choice is matter of preference -- ultimately both choices can work.

@YohannDudouit:
> I really don't understand why you don't want to make the code safer and cleaner when this has no performance impact.

I prefer, for this low-level class, the current, more explicit design where all basic operation like initialization, deletion, etc are done explicitly similar to how one works with pointers. For this class, I think this design actually makes the code easier to understand and follow. The concern that this is bug-prone is not not a big concern for me since this is a low-level, internally used class.

I'm just expressing my opinion here. If the majority prefers to accept this PR, I'm perfectly fine with that.

**Comment by tzanio** [comment]
I don't have a strong preference either way. On one hand this is a low-level concept and using the corresponding low-level abstraction (raw pointer) makes sense. On the other hand, `Memory` is a class, and so I can one can expect it to behave in a class-like way. 

Maybe it will be best if we ask the folks that may work with this class more directly what they think -- @camierjs, @artv3, @tomstitt, @dylan-copeland -- do you have a preference one way or the other?

**Comment by dylan-copeland** [comment]
The arguments on both sides make sense, and I doubt it will make a big difference, but my personal preference is to be cautious and initialize things rather than relying on human understanding of the correct behavior, although that understanding may be necessary anyway. So I slightly prefer to use this PR.

**Review by acfisher — COMMENTED** [review]
Given that it will be inconsistent with the semantics of raw pointers either way, I vote for not having to litter our code with var.Reset() everywhere.  Since this is done for us it will be one fewer way to shoot ourselves in the foot with the memory manager.

**Review by acfisher — APPROVED** [review]
This has been sitting here for a while I think we should either pull it in or deny it and move on.  I vote for pulling it in.  So I'll approve it.

**Comment by jamiebramwell** [comment]
As a user, I think this is a large improvement and I strongly support this getting merged. Uninitialized pointer variable errors are very hard to track down.

**Review by pazner — APPROVED** [review]

**Review by camierjs — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**
